File: ai/fixtures.py
"""Development fixtures, loaded from `data/`.

The Node backend reads the exact same files (see `backend/src/utils/fixtures.ts`),
so a fixture is a shared artefact rather than two copies that drift apart.

Anything served from here is marked `mock=True` all the way out to the browser.
"""

from __future__ import annotations

import json
from functools import lru_cache
from typing import Any

from config import EVALUATION_DIR, MOCK_DIR
from schemas import (
    AudienceGraph,
    ContentDNA,
    CounterfactualExperiment,
    EvaluationDataset,
    Persona,
    RunMetadata,
    SimulationResult,
)


@lru_cache(maxsize=None)
def _read(directory_key: str, filename: str) -> str:
    directory = MOCK_DIR if directory_key == "mock" else EVALUATION_DIR
    path = directory / filename
    if not path.is_file():
        raise FileNotFoundError(
            f"Fixture '{filename}' not found at {path}. "
            "Fixtures live in data/ — see data/README.md."
        )
    return path.read_text(encoding="utf-8")


def _load(directory_key: str, filename: str) -> Any:
    return json.loads(_read(directory_key, filename))


def audience_graph() -> AudienceGraph:
    return AudienceGraph.model_validate(_load("mock", "audience_graph.json"))


def personas() -> list[Persona]:
    return [Persona.model_validate(item) for item in _load("mock", "personas.json")]


def personas_for_segment(segment_id: str, count: int | None = None) -> list[Persona]:
    matches = [persona for persona in personas() if persona.segment_id == segment_id]
    # Fall back to the full pool so a newly invented segment id still gets
    # something to work with rather than an empty simulation.
    pool = matches or personas()
    return pool[:count] if count else pool


def content_dna() -> ContentDNA:
    return ContentDNA.model_validate(_load("mock", "content_dna.json"))


def simulation_result() -> SimulationResult:
    return SimulationResult.model_validate(_load("mock", "simulation_result.json"))


def counterfactual_experiment() -> CounterfactualExperiment:
    return CounterfactualExperiment.model_validate(
        _load("mock", "counterfactual_experiment.json")
    )


def evaluation_dataset() -> EvaluationDataset:
    return EvaluationDataset.model_validate(_load("evaluation", "historical_reels.json"))


def mock_metadata(**overrides: Any) -> RunMetadata:
    """Metadata stamped on every fixture-backed response."""
    base = {"model": "fixture", "prompt_version": "n/a", "latency_ms": 0.0, "mock": True}
    base.update(overrides)
    return RunMetadata.model_validate(base)

File: backend/src/utils/fixtures.ts
import fs from 'node:fs';
import path from 'node:path';

import { MOCK_DIR } from '../config/paths';
import { ApiError } from './ApiError';
import { logger } from './logger';

/**
 * Load development fixtures from `data/mock_personas/`.
 *
 * Fixtures live on disk rather than inline in TypeScript so the Python AI
 * service and the Node backend read the exact same bytes — see data/README.md.
 * Results are cached because these files never change while the process runs.
 */
const cache = new Map<string, unknown>();

export function loadFixture<T>(filename: string): T {
  const cached = cache.get(filename);
  if (cached !== undefined) return cached as T;

  const filePath = path.join(MOCK_DIR, filename);

  try {
    const parsed = JSON.parse(fs.readFileSync(filePath, 'utf-8')) as T;
    cache.set(filename, parsed);
    return parsed;
  } catch (error) {
    logger.error('fixture_load_failed', { filePath, error: String(error) });
    throw new ApiError(
      'INTERNAL',
      `Could not load development fixture '${filename}'. Expected it at ${filePath}.`,
    );
  }
}

export const FIXTURES = {
  audienceGraph: 'audience_graph.json',
  personas: 'personas.json',
  contentDna: 'content_dna.json',
  simulationResult: 'simulation_result.json',
  counterfactualExperiment: 'counterfactual_experiment.json',
} as const;
